src/system/OS_info/os_info_windows.cpp

```diff
@@ -80,10 +83,38 @@ static std::string version_name() {
 	return ret.substr(0, ret.find('|'));
 }
 
+unsigned int build_number() {
+	HKEY hkey{};
+	if(RegOpenKeyExA(HKEY_LOCAL_MACHINE, R"(Software\Microsoft\Windows NT\CurrentVersion)", 0, KEY_READ, &hkey))
+		return {};
+
+	DWORD ubr      = 0;
+	DWORD ubr_size = sizeof(ubr);
+	if(!RegQueryValueExA(hkey, "UBR", nullptr, nullptr, reinterpret_cast<LPBYTE>(&ubr), &ubr_size))
+		return ubr;
 
+	// Fall back to BuildLabEx in the early version of Windows 8.1 and less.
+	DWORD buildlabex_size = 0;
+	if(RegQueryValueExA(hkey, "BuildLabEx", nullptr, nullptr, nullptr, &buildlabex_size))
+		return {};
+
+	std::vector<char> buildlabex(buildlabex_size);
+	if(RegQueryValueExA(hkey, "BuildLabEx", nullptr, nullptr, reinterpret_cast<LPBYTE>(buildlabex.data()), &buildlabex_size))
+		return {};
+
+	auto token = std::strtok(buildlabex.data(), ".");
```

Comment:
`strtok_r`

Reply:
It seems `strtok_r` is not in standard too, would there be any other alternative? I am concerned about the non-thread-safety of (almost of)C standard functions, including the `strtok` family...

Reply:
strtok_r is part of plain XPG/SUS/POSIX since Issue 5 (SUSv2, 1997), imported from POSIX.1c-1995 (Threads Extension)

but the Microsoft libc actually does not seem to have it, since NT 3.1 predates the Threads Extension, but does have [`strtok_s`](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strtok-s-strtok-s-l-wcstok-s-wcstok-s-l-mbstok-s-mbstok-s-l?view=msvc-160) with the same semantics (which is fittingly inane) – `#define strtok_r(...) strtok_s(__VA_ARGS__)` when building for MSlibc should work?

Reply:
`strtok_s` of MS CRT(which differs from `strtok_s` in C11) and `strtok_r` of POSIX receive the same arguments, so
```C++
#ifdef _MSC_VER
#define strtok_r strtok_s
#endif
```
works well. But I begin to think again `std::string::find` will be more safe and straightforward... ;)